snapcraft/internal/remote_build/_worktree.py

```diff
@@ -14,100 +14,263 @@
-    is maintained under revision control.
-    """
+    def __init__(
+        self, worktree_dir: str, project: Project, package_all_sources=False
+    ) -> None:
+        self._project = project
+        self._package_all_sources = package_all_sources
 
-    def __init__(self, src: str, dest: str, ignore: List[str] = []) -> None:
-        if not os.path.isdir(src):
-            raise RuntimeError
-        self._srcroot = src
-        self._destroot = dest
-        self._ignore = [".git", ".gitignore", ".gitmodules", ".bzr", ".svn"]
-        self._ignore.extend(ignore)
-        Path(dest).mkdir(parents=True, exist_ok=True)
-        self._repo = Repo(dest)
-        self._repo.reset()
+        # Get snapcraft yaml (as OrderedDict).
+        self._snapcraft_yaml = (
+            self._project.info.get_raw_snapcraft()
+        )  # type: OrderedDict
 
-    def sync(self):
-        """
-        Synchronize the source directory and the work tree.
+        # Working copy.
+        self._prepared_snapcraft_yaml = self._project.info.get_raw_snapcraft()
 
-        Given the path for a source tree, replicate its contents and keep the destination
-        in a revision control repository.
-        """
+        # Working tree base directory.
+        self._base_dir = worktree_dir
+
+        # Initialize cache for sources.
+        self._cache_dir = os.path.join(self._base_dir, "cache")
+        os.makedirs(self._cache_dir, exist_ok=True)
+
+        # Initialize clean repo to ship to remote builder.
+        self._repo_dir = os.path.join(self._base_dir, "repo")
+        if os.path.exists(self._repo_dir):
+            shutil.rmtree(self._repo_dir)
+
+        self._repo_sources_dir = os.path.join(self._repo_dir, "sources")
+        self._repo_snap_dir = os.path.join(self._repo_dir, "snap")
+
+    def _get_part_source_handler(self, part_name: str, source: str, source_dir: str):
+        # The actual source may be a selector, override it.
+        part = self._snapcraft_yaml["parts"][part_name].copy()
+        part["source"] = source
+        source_type = part.get("source-type", None)
+
+        handler_class = snapcraft.internal.sources.get_source_handler(
+            source, source_type=source_type
+        )
+        return handler_class(
+            source,
+            source_dir,
+            source_checksum=part.get("source-checksum", None),
+            source_branch=part.get("source-branch", None),
+            source_tag=part.get("source-tag", None),
+            source_depth=part.get("source-depth", None),
+            source_commit=part.get("source-commit", None),
+        )
+
+    def _printable_part_name(self, part_name: str, selector=None) -> str:
+        if selector:
+            return "{}[{}]".format(part_name, selector)
+        return part_name
+
+    def _get_part_cache_dir(self, part_name: str, selector=None) -> str:
+        """Get source cache directory for part_name."""
```

Comment:
can you add a docstring for what the selector is?

Reply:
> Can `add --package-all-sources to pull and package all source types for the builder.` be a separate commit?



> Can `add --package-all-sources to pull and package all source types for the builder.` be a separate commit?

Will do.